<a href="https://colab.research.google.com/github/michalis0/MGT-502-Data-Science-and-Machine-Learning/blob/main/assignments/Part 2/Assignment_part_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## DSML investigation

You are part of the Suisse Impossible Mission Force, or SIMF for short. You need to uncover a rogue agent that is trying to steal sensitive information.

Your mission, should you choose to accept it, is to find that agent before stealing any classified information. Good luck!

# Assignment part two

We received new intelligence about the rogue agent which lets us discover his age.
Use the table "social_media_SIMF_archive.csv" to create a model that predicts the age of a person. Use this model to predict the ages of the table of suspects "suspect_social_media_statistics.csv".



## Getting to know our data

- userId: Unique number given to each user.

- tier: Tier of the city in which the user is residing.

- gender: Categorical feature representing the gender of the user. 1 represents male and 2 represents female.

- following_rate: Number of accounts followed by the user(feature is normalized)

- followers_avg_age: Average of age groups of all the followers of the user.

- following_avg_age: Average of age groups of all the accounts followed by the user.

- max_repetitive_punc: Maximum repititive punctuations found in the bio and comments of the user.

- num_of_hashtags_per_action: Average nubmer of hashtags used by the user per comment.

- emoji_count_per_action: Average number of emojis used by the user per comment

- punctuations_per_action: Average number of punctuations used by the user per comment.

- number_of_words_per_action: Average number of words used by the user per comment.

- avgCompletion: Average watch time completion rate of the videos.

- avgTimeSpent: Average time spent by the user on a video in seconds.

- avgDuration: Average duration of the videos that the user has watched till date.

- avgComments: Average number of comments per video watched.

- creations: Total number of videos uploaded by the user.

- content_views: Total number of videos watched.

- num_of_comments: Total number of comments made by the user (normalized)

- weekends_trails_watched_per_day: Number of videos watched on weekends per day.

- weekdays_trails_watched_per_day: Number of videos watched on weekdays per day.

- slot1_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot2_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot3_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot4_trails_watched_per_day: data

- avgt2: Average number of followers of all the accounts followed by the user.

- age : The age of the user


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler


train_dataset = pd.read_csv("https://raw.githubusercontent.com/michalis0/MGT-502-Data-Science-and-Machine-Learning/refs/heads/main/assignments/Part%202/data/social_media_SIMF_archive.csv")
suspects = pd.read_csv("https://raw.githubusercontent.com/michalis0/MGT-502-Data-Science-and-Machine-Learning/refs/heads/main/assignments/Part%202/data/suspects_social_media.csv")

In [2]:
# Set the random seed for reproducibility
np.random.seed(42)

### Training a model
Select the features which by convention are called X. We will also choose the target variable which we typically call y.


In [3]:
FEATURES_COLS = [
    'tier', 
    'gender', 
    'following_rate', 
    'followers_avg_age', 
    'following_avg_age', 
    'max_repetitive_punc', 
    'num_of_hashtags_per_action', 
    'emoji_count_per_action', 
    'punctuations_per_action', 
    'number_of_words_per_action', 
    'avgCompletion', 
    'avgTimeSpent', 
    'avgDuration', 
    'avgComments', 
    'creations', 
    'content_views', 
    'num_of_comments', 
    'weekends_trails_watched_per_day', 
    'weekdays_trails_watched_per_day', 
    'slot1_trails_watched_per_day', 
    'slot2_trails_watched_per_day', 
    'slot3_trails_watched_per_day', 
    'slot4_trails_watched_per_day', 
    'avgt2'
]

LABEL_COL = 'age'

X = train_dataset[FEATURES_COLS]
y = train_dataset[LABEL_COL]


Now do the train-test split, using 80% of the data as training and 20% as test. Set the random state to 42.

In [4]:
#train test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, shuffle=True)

Next, rescale the features in training and test data using `MinMaxScaler`.

In [5]:
# Use the MinMaxScaler to rescale your the features in X_train and X_test
scaler = MinMaxScaler()

#fit and transform the training data
X_train = scaler.fit_transform(X_train)
#transform the test data
X_test = scaler.transform(X_test)

## Linear regression
Now, create a linear regression model. Train it on the training data and use the model to make predictions on the test data. How is the performance of your model on the test data?

*Note: For the liner regression model, use the default attribute value as seen in the Lab.*

In [6]:
# YOUR CODE HERE
# do the right imports
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score # Metrics for errors
# create the model

model = LinearRegression()

# Fit the model
model.fit(X_train, y_train)


LinearRegression()

and we do the predictions on the test set:

In [7]:
# YOUR CODE HERE
# Predict the test data
predictions = model.predict(X_test)

What are the model's $MAE$, $MSE$ and $R^2$ ?

In [8]:
# YOUR CODE HERE
# Compute the MAE, the MSE and the R^2 on the test set
mae = mean_absolute_error(y_test, predictions)
mse = mean_squared_error(y_test, predictions)
r2 = r2_score(y_test, predictions)

# Compute the MAE, the MSE and the R^2 on the training set
predictions_train = model.predict(X_train)
mae_train = mean_absolute_error(y_train, predictions_train)
mse_train = mean_squared_error(y_train, predictions_train)
r2_train = r2_score(y_train, predictions_train)

print(f"MAE test set: {mae:0.2f}; MAE training set: {mae_train:0.2f};")
print(f"MSE test set: {mse:0.2f}; MSE training set: {mse_train:0.2f};")
print(f"R\u00b2 test set: {r2:0.2f}; R\u00b2 training set: {r2_train:0.2f};" )

MAE test set: 5.33; MAE training set: 5.37;
MSE test set: 48.95; MSE training set: 49.40;
R² test set: 0.66; R² training set: 0.66;


**Q1. What is the mean absolute error (MAE) of the model over the test set?**

*Hint : The correct value should be between 5.20 and 5.50*

Round to 2 digits (e. g. 4.32)

5.33

**Q2. What is the mean squared error (MSE) of the model over the test set?"**

*Hint : The correct value should be between 48 and 50*

Round to 2 digit (e.g. 4.32)

48.95

**Q3. What is the $R^2$ score of the model over the test set?**

*Hint : The correct value should be between 0.55 and 0.75*

Round to 2 digits (e. g. 4.32)

0.66

## Retraining on the full dataset
Training on a split of the training dataset, as you just did, is mainly for evaluating the performance of the resulting model on the unseen data and also optimizing various hyper-parameters of the model.

We can now retrain the model on the full training data prior to making predictions.

Before predicting the age of the suspects, retrain the model on the full training dataset.

You can reuse the same features as before, but make sure those are also available on the suspects dataset.

In [9]:
# YOUR CODE HERE
X_full = train_dataset[FEATURES_COLS]
y_full = train_dataset[LABEL_COL]
display(X_full)
display(suspects)

tier  gender  following_rate  followers_avg_age  following_avg_age  \
0          2       1        0.000000               0.00           0.000000   
1          3       1        0.011096               3.00           3.000000   
2          2       2        0.064516               1.40           2.142857   
3          2       1        0.000000               0.00           0.000000   
4          2       1        0.357143               3.00           2.000000   
...      ...     ...             ...                ...                ...   
486044     3       1        0.001923               0.00           0.000000   
486045     2       2        0.000000               0.00           0.000000   
486046     2       1        0.766129               1.25           2.454545   
486047     3       1        0.000000               0.00           0.000000   
486048     2       1        0.000000               0.00           0.000000   

        max_repetitive_punc  num_of_hashtags_per_action  \
0                         0                         0.0   
1                         0                         0.0   
2                         0                         0.0   
3                         0                         0.0   
4                         7                         0.0   
...                     ...                         ...   
486044                    2                         0.0   
486045                    0                         0.0   
486046                    0                         0.0   
486047                    0                         0.0   
486048                    0                         0.0   

        emoji_count_per_action  punctuations_per_action  \
0                          0.0                 0.000000   
1                          0.0                 0.000000   
2                          0.0                 0.133333   
3                          0.0                 0.000000   
4                          0.0                 0.000000   
...                        ...                      ...   
486044                     0.0                 1.571429   
486045                     0.0                 0.000000   
486046                     0.0                 0.000000   
486047                     0.0                 0.000000   
486048                     0.0                 0.000000   

        number_of_words_per_action  ...  creations  content_views  \
0                             0.00  ...   0.000000       0.190083   
1                             0.90  ...   0.005548       0.020804   
2                             0.40  ...   0.008065       0.104839   
3                             0.00  ...   0.000000       0.007937   
4                             0.05  ...   0.010204       0.765306   
...                            ...  ...        ...            ...   
486044                        9.00  ...   0.005769       0.005769   
486045                        0.00  ...   0.000000       0.194444   
486046                        0.08  ...   0.008065       0.177419   
486047                        0.00  ...   0.000000       0.015873   
486048                        0.00  ...   0.000000       0.061224   

        num_of_comments  weekends_trails_watched_per_day  \
0              0.000000                         0.000000   
1              0.000000                         0.004854   
2              0.000000                         0.060484   
3              0.000000                         0.000000   
4              0.030612                         0.000000   
...                 ...                              ...   
486044         0.000000                         0.000000   
486045         0.000000                         0.000000   
486046         0.008065                         0.076613   
486047         0.000000                         0.000000   
486048         0.000000                         0.000000   

        weekdays_trails_watched_per_day  slot1_trails_watched_per_day  \
0                             

Unnamed: 0  userId  tier  gender  following_rate  followers_avg_age  \
0             0  317991     2       1        0.000000           0.000000   
1             1  241892     2       2        0.000000           0.000000   
2             2  303376     2       1        0.160000           1.529412   
3             3  761992     2       2        0.032000           1.000000   
4             4  373318     2       1        0.009524           0.000000   
..          ...     ...   ...     ...             ...                ...   
724         709  458293     2       2        0.065421           0.000000   
725         710  218415     2       1        0.000000           0.000000   
726         711  173906     2       1        0.000000           0.000000   
727         712  178685     2       2        0.000000           0.000000   
728         713  200865     2       1        0.057377           1.500000   

     following_avg_age  max_repetitive_punc  num_of_hashtags_per_action  \
0             0.000000                    0                         0.0   
1             0.000000                    0                         0.0   
2             2.176471                    0                         0.0   
3             2.500000                    0                         0.0   
4             0.000000                    3                         0.0   
..                 ...                  ...                         ...   
724           0.000000                    3                         0.0   
725           0.000000                    0                         0.0   
726           0.000000                    0                         0.0   
727           0.000000                    0                         0.0   
728           2.750000                    0                         0.0   

     emoji_count_per_action  ...  creations  content_views  num_of_comments  \
0                       0.0  ...   0.000000       0.123967            0.000   
1                       0.0  ...   0.000000       0.031746            0.000   
2                       0.0  ...   0.008000       0.520000            0.000   
3                       0.0  ...   0.008000       0.360000            0.008   
4                       0.0  ...   0.047619       0.552381            0.000   
..                      ...  ...        ...            ...              ...   
724                     0.0  ...   0.009346       0.672897            0.000   
725                     0.0  ...   0.008264       0.173554            0.000   
726                     0.0  ...   0.000000       0.041322            0.000   
727                     0.0  ...   0.000000       0.149533            0.000   
728                     0.0  ...   0.024590       0.213115            0.000   

     weekends_trails_watched_per_day  weekdays_trails_watched_per_day  \
0                           0.000000                         0.033058   
1                           0.000000                         0.006349   
2                           0.248000                         0.009600   
3                           0.224000                         0.001600   
4                           0.047619                         0.139048   
..                               ...                              ...   
724                         0.000000                         0.181308   
725                         0.000000                         0.036364   
726                         0.000000                         0.009917   
727                         0.000000                         0.028037   
728                         0.000000                         0.072131   

     slot1_trails_watched_per_day  slot2_trails_watched_per_day  \
0                        0.165289                      0.000000   
1                        0.000000                      0.000000   
2                        0.000000                      0.048000   
3                        0.000000                      0.008000   
4                        0.000

In [10]:
# YOUR CODE HERE
# Use a new MinMaxScaller to rescale your the features in X and in the suspect dataset
scaler = MinMaxScaler()
X_full = scaler.fit_transform(X_full)
X_suspects = scaler.transform(suspects[FEATURES_COLS])

In [11]:
# YOUR CODE HERE
# create the new model
model = LinearRegression()

# Fit the new model
model.fit(X_full, y_full)

LinearRegression()

In [12]:
# YOUR CODE HERE
# Make the predictions
predictions = model.predict(X_suspects)

**Keep only the suspects between 25 and 35 years old. Use the `age` values you predicted in the previous step.**


In [13]:
# YOUR CODE HERE
# Filter suspects based on the predicted ages
suspects_filtered = suspects[(predictions >= 25) & (predictions <= 35)]
suspects_filtered

Unnamed: 0  userId  tier  gender  following_rate  followers_avg_age  \
75           75  785994     2       2         0.00000                0.0   
77           75  123699     2       2         0.00000                0.0   
78           75  139646     2       2         0.00000                0.0   
79           75  255830     2       2         0.00000                0.0   
80           75  267733     2       2         0.00000                0.0   
81           75  710764     2       2         0.00000                0.0   
82           75  297653     2       2         0.00000                0.0   
83           75  410319     2       2         0.00000                0.0   
84           75  411537     2       2         0.00000                0.0   
85           75  424791     2       2         0.00000                0.0   
86           75  428951     2       2         0.00000                0.0   
87           75  638911     2       2         0.00000                0.0   
88           75  628854     2       2         0.00000                0.0   
89           75  649131     2       2         0.00000                0.0   
90           75  861915     2       2         0.00000                0.0   
91           75  942571     2       2         0.00000                0.0   
114          99  538431     3       2         0.00813                0.0   
187         172  822072     3       1         0.00000                0.0   
431         416  520750     2       2         0.00000                0.0   
515         500  773146     2       2         0.00000                0.0   
601         586  444557     2       2         0.00000                0.0   
683         668  241540     2       2         0.00000                0.0   

     following_avg_age  max_repetitive_punc  num_of_hashtags_per_action  \
75                 0.0                    0                         0.0   
77                 0.0                    0                         0.0   
78                 0.0                    0                         0.0   
79                 0.0                    0                         0.0   
80                 0.0                    0                         0.0   
81                 0.0                    0                         0.0   
82                 0.0                    0                         0.0   
83                 0.0                    0                         0.0   
84                 0.0                    0                         0.0   
85                 0.0                    0                         0.0   
86                 0.0                    0                         0.0   
87                 0.0                    0                         0.0   
88                 0.0                    0                         0.0   
89                 0.0                    0                         0.0   
90                 0.0                    0                         0.0   
91                 0.0                    0                         0.0   
114                0.0                    0                         0.0   
187                0.0                    0                         0.0   
431                0.0                    0                         0.0   
515                0.0                    0                         0.0   
601                0.0                    0                         0.0   
683                0.0                    0                         0.0   

     emoji_count_per_action  ...  creations  content_views  num_of_comments  \
75                      0.0  ...    0.00000       0.008264              0.0   
77                      0.0  ...    0.00000       0.008264              0.0   
78                      0.0  ...    0.00000       0.008264              0.0   
79                      0.0  ...    0.00000       0.008264              0.0   
80                      0.0  ...    0.00000       0.008264              0.0   
81                      0.0  ...    0.00000       0.0082

**Q4. Which of these users remain suspects after filtering for those aged between 25 and 35?** (check all that apply)

785994,
123699,
410319,
255830

The SIMF is considering narrowing down the suspect pool by focusing on gender. As a preliminary step, they need to know the percentage of female suspects within the age range of 25 to 35 years old.

Calculate and report the percentage of female suspects among those aged 25 to 35 years old from the suspects you identified previously.

*Hint: In the provided dataset, 'gender' is coded with 1 representing male users and 2 representing female users.*

In [14]:
# YOUR CODE HERE
# Calculate the percentage of female suspects
female_suspects = suspects_filtered[suspects_filtered['gender'] == 2]
percentage_female = (len(female_suspects) / len(suspects_filtered)) * 100
print(f"{percentage_female:0.2f}")

95.45


**Q5. What is the percentage of female users among the predicted suspects aged between 25 and 35 years old?**

Round to 2 digits (e. g. 4.32)



95.45

**What is the predicted age of a Female user with the following characteristics:**
* tier = 2
* following_rate = 2
* followers_avg_age = 27
* following_avg_age = 22
* max_repetitive_punc = 0.06
* num_of_hashtags_per_action = 0.8
* emoji_count_per_action = 0.3
* punctuations_per_action = 0.03
* number_of_words_per_action = 7
* avgCompletion = 12
* avgTimeSpent = 0.13
* avgDuration = 0.8
* avgComments = 0.04
* likeToDislikeRation = 0.76
* creations = 10
* content_views = 65
* num_of_comments = 12
* weekends_trails_watched_per_day = 0
* weekdays_trails_watched_per_day = 0.00353
* slot1_trails_watched_per_day = 0.032
* slot2_trails_watched_per_day = 0
* slot3_trails_watched_per_day = 0.004
* slot4_trails_watched_per_day = 2
* avgt2 = 39

*hint: make sure the features and their order corresponds to the ones used for the training.*

*hint 2: don't forget to reprocess the data as you did before*

In [15]:
user_details = {
    "tier": [2],
    "gender": [2],
    "following_rate": [2],
    "followers_avg_age": [27],
    "following_avg_age": [22],
    "max_repetitive_punc": [0.06],
    "num_of_hashtags_per_action": [0.8],
    "emoji_count_per_action": [0.3],
    "punctuations_per_action": [0.03],
    "number_of_words_per_action": [7],
    "avgCompletion": [12],
    "avgTimeSpent": [0.13],
    "avgDuration": [0.8],
    "avgComments": [0.04],
    "creations": [10],
    "content_views": [65],
    "num_of_comments": [12],
    "weekends_trails_watched_per_day": [0],
    "weekdays_trails_watched_per_day": [0.00353],
    "slot1_trails_watched_per_day": [0.032],
    "slot2_trails_watched_per_day": [0],
    "slot3_trails_watched_per_day": [0.004],
    "slot4_trails_watched_per_day": [2],
    "avgt2": [39]
}

# YOUR CODE HERE
user_df = pd.DataFrame(user_details)
X_new_user = scaler.transform(user_df)

prediction = model.predict(X_new_user)

print(f"Age prediction: {prediction[0]:.0f}")


Age prediction: 26


**Q6. What is the predicted age of a female user with those characteristics using the regression model that you created?**

Enter the rounded value as an integer (e. g. 21 if your result is 21.34 or 20.79)

26

## Your investigation is progressing effectively, and the list of suspects is narrowing down.

**Don't forget to answer the quiz and submit your code on Moodle before the end of the deadline.**